diff --git a/Install b/Install
index 7bd0f77..8d28cd9 100755
--- a/Install
+++ b/Install
@@ -10,14 +10,14 @@ TOP=`pwd`
 
 # if no Configuration file, bail out
 
-if [ ! -f ${TOP}/Configuration ]; then
+if [ ! -f "${TOP}/Configuration" ]; then
 	echo "Missing Configuration file"
 	exit 1
 fi
 
 # source the Configuration file
 
-. ${TOP}/Configuration
+. "${TOP}/Configuration"
 echo "Sourced Configuration file"
 
 # add/update links to config.py library
@@ -26,19 +26,19 @@ CONFIG=${TOP}/lib/python/config.py
 
 for dir in gather control schema
 do
-	if [ -h ${TOP}/${dir}/config.py ]; then
-		rm ${TOP}/${dir}/config.py
+	if [ -h "${TOP}/${dir}/config.py" ]; then
+		rm "${TOP}/${dir}/config.py"
 	fi
-	if [ -f ${TOP}/${dir}/config.pyc ]; then
-		rm ${TOP}/${dir}/config.pyc
+	if [ -f "${TOP}/${dir}/config.pyc" ]; then
+		rm "${TOP}/${dir}/config.pyc"
 	fi
-	ln -s ${CONFIG} ${TOP}/${dir}/config.py 
+	ln -s "${CONFIG}" "${TOP}/${dir}/config.py"
 	echo "Created symlink: ${TOP}/${dir}/config.py"
 done
 
 # back to top level...
 
-cd ${TOP}
+cd "${TOP}"
 
 # if we don't have a data and a logs directory, make them
 
@@ -54,7 +54,7 @@ fi
 
 # update permissions on gatherers and controllers
 
-chmod ug+x ${TOP}/gather/*Gatherer.py ${TOP}/control/*
+chmod ug+x "${TOP}/gather/"*Gatherer.py "${TOP}/control/"*
 echo "Granted execute permissions for gatheres and control scripts"
 
 echo "Finished"
